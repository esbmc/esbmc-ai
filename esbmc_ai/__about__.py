# Author: Yiannis Charalambous

__version__ = "v0.7.0"
__author__: str = "Yiannis Charalambous"
