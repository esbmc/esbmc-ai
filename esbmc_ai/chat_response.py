# Author: Yiannis Charalambous

from enum import Enum
from typing import NamedTuple

from langchain.schema import (
    AIMessage,
    BaseMessage,
    HumanMessage,
    SystemMessage,
)


"""Contains classes and functions that relate to sending/receiving messages from LLMs."""


class FinishReason(Enum):
    # API response still in progress or incomplete
    null = 0
    # API returned complete model output
    stop = 1
    # Incomplete model output due to max_tokens parameter or token limit
    length = 2
    # Omitted content due to a flag from our content filters
    content_filter = 3


class ChatResponse(NamedTuple):
    message: BaseMessage = AIMessage(content="")
    total_tokens: int = 0
    finish_reason: FinishReason = FinishReason.null


def dict_to_base_message(json_string: dict) -> BaseMessage:
    """Converts a json representation of messages (such as in config.json),
    into LangChain object messages. The three recognized roles are:
    1. System
    2. AI
    3. Human"""
    role: str = json_string["role"]
    content: str = json_string["content"]
    if role == "System":
        return SystemMessage(content=content)
    elif role == "AI":
        return AIMessage(content=content)
    elif role == "Human":
        return HumanMessage(content=content)
    else:
        raise Exception()


def list_to_base_messages(json_messages: list[dict]) -> tuple[BaseMessage, ...]:
    """Converts a list of messages from JSON format to a list of BaseMessage."""
    return tuple(dict_to_base_message(msg) for msg in json_messages)
