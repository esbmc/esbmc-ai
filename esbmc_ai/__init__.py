"""ESBMC-AI is an AI augmentation layer over ESBMC, the Efficient
Software Bounded Model Checker. With the power of LLMs, it adds
features such as automatic code fixing and more."""

from esbmc_ai.__about__ import __version__, __author__

from esbmc_ai.config import Config
from esbmc_ai.chat_command import ChatCommand

__all__ = ["Config", "ChatCommand"]
