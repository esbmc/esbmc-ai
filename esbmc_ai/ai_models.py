# Author: Yiannis Charalambous

from abc import abstractmethod
from typing import Any, Iterable, Optional, Union
from enum import Enum
from langchain_core.language_models import BaseChatModel
from pydantic.v1.types import SecretStr
from typing_extensions import override

from langchain_openai import ChatOpenAI
from langchain_ollama import ChatOllama

from langchain.prompts.chat import ChatPromptTemplate
from langchain.schema import (
    BaseMessage,
    PromptValue,
)


from esbmc_ai.api_key_collection import APIKeyCollection


class AIModel(object):
    """This base class represents an abstract AI model."""

    name: str
    tokens: int

    def __init__(
        self,
        name: str,
        tokens: int,
    ) -> None:
        self.name = name
        self.tokens = tokens

    @abstractmethod
    def create_llm(
        self,
        api_keys: APIKeyCollection,
        temperature: float = 1.0,
        requests_max_tries: int = 5,
        requests_timeout: float = 60,
    ) -> BaseChatModel:
        """Initializes a large language model model with the provided parameters."""
        raise NotImplementedError()

    @classmethod
    def convert_messages_to_tuples(
        cls, messages: Iterable[BaseMessage]
    ) -> list[tuple[str, str]]:
        """Converts messages into a format understood by the ChatPromptTemplate - since it won't format
        BaseMessage derived classes for some reason, but will for tuples, because they get converted into
        Templates in function `_convert_to_message`."""
        return [(message.type, str(message.content)) for message in messages]

    @classmethod
    def escape_messages(
        cls,
        messages: Iterable[BaseMessage],
        allowed_keys: list[str],
    ) -> Iterable[BaseMessage]:
        """Adds escape curly braces to the messages, will make sure that the sequential
        curly braces in the messages is even (and hence escaped). Will ignore curly braces
        with `allowed_keys`."""

        def add_safeguards(content: str, char: str, allowed_keys: list[str]) -> str:
            start_idx: int = 0
            while True:
                char_idx: int = content.find(char, start_idx)
                if char_idx == -1:
                    break
                # Count how many sequences of the char are occuring
                count: int = 1
                # FIXME Add bounds check here
                while (
                    len(content) > char_idx + count
                    and content[char_idx + count] == char
                ):
                    count += 1

                # Check if the next sequence is in the allowed keys, if it is, then
                # skip to the next one.
                is_allowed: bool = False
                for key in allowed_keys:
                    if key == content[char_idx + count : char_idx + count + len(key)]:
                        is_allowed = True
                        break

                # Now change start_idx to reflect new location. The start index is at the end of
                # all the chars (including the inserted one).
                start_idx = char_idx + count + 1

                # If inside allowed keys, then continue to next iteration.
                if is_allowed:
                    continue

                # Check if odd number (will need to add extra char)
                if count % 2 != 0:
                    content = (
                        content[: char_idx + count] + char + content[char_idx + count :]
                    )
            return content

        reversed_keys: list[str] = [key[::-1] for key in allowed_keys]

        result: list[BaseMessage] = []
        for msg in messages:
            content: str = str(msg.content)
            look_pointer: int = 0
            # Open curly check
            if content.find("{", look_pointer) != -1:
                content = add_safeguards(content, "{", allowed_keys)
            # Close curly check
            if content.find("}", look_pointer) != -1:
                # Do it in reverse with reverse keys.
                content = add_safeguards(content[::-1], "}", reversed_keys)[::-1]
            new_msg = msg.copy()
            new_msg.content = content
            result.append(new_msg)
        return result

    def apply_chat_template(
        self,
        messages: Iterable[BaseMessage],
        **format_values: Any,
    ) -> PromptValue:
        """Applies the formatted values onto the message chat template. For example,
        if the message contains the token {source}, then format_values contains a
        value for {source} then it will be substituted."""
        escaped_messages = AIModel.escape_messages(messages, list(format_values.keys()))
        message_tuples = AIModel.convert_messages_to_tuples(escaped_messages)
        return ChatPromptTemplate.from_messages(messages=message_tuples).format_prompt(
            **format_values,
        )


class AIModelOpenAI(AIModel):
    context_length_exceeded_error: str = "context_length_exceeded"
    """Error code for when the length has been reached."""

    @override
    def create_llm(
        self,
        api_keys: APIKeyCollection,
        temperature: float = 1.0,
        requests_max_tries: int = 5,
        requests_timeout: float = 60,
    ) -> BaseChatModel:
        assert api_keys.openai, "No OpenAI api key has been specified..."
        return ChatOpenAI(
            model=self.name,
            api_key=SecretStr(api_keys.openai),
            max_tokens=None,
            temperature=temperature,
            max_retries=requests_max_tries,
            timeout=requests_timeout,
            model_kwargs={},
        )

class OllamaAIModel(AIModel):
    def __init__(self, name: str, tokens: int, url: str) -> None:
        super().__init__(name, tokens)
        self.url: str = url
    
    @override
    def create_llm(self, api_keys: APIKeyCollection, temperature: float = 1, requests_max_tries: int = 5, requests_timeout: float = 60) -> BaseChatModel:
        # Ollama does not use API keys
        _ = api_keys
        _ = requests_max_tries
        return ChatOllama(
            base_url=self.url,
            model=self.name,
            temperature=temperature,
            client_kwargs={
                "timeout":requests_timeout,
            },
        )


class _AIModels(Enum):
    """Private enum that contains predefined AI Models. OpenAI models are not
    defined because they are fetched from the API."""

    # FALCON_7B = OllamaAIModel(...)
    pass


_custom_ai_models: list[AIModel] = []

_ai_model_names: set[str] = set(item.value.name for item in _AIModels)


def add_custom_ai_model(ai_model: AIModel) -> None:
    """Registers a custom AI model."""
    # Check if AI already already exists.
    if ai_model.name in _ai_model_names:
        raise Exception(f'AI Model "{ai_model.name}" already exists...')
    _ai_model_names.add(ai_model.name)
    _custom_ai_models.append(ai_model)


def is_valid_ai_model(
    ai_model: Union[str, AIModel], api_keys: Optional[APIKeyCollection] = None
) -> bool:
    """Accepts both the AIModel object and the name as parameter. It checks the
    openai servers to see if a model is defined on their servers, if not, then
    it checks the internally defined AI models list."""

    # Get the name of the model
    name: str = ai_model.name if isinstance(ai_model, AIModel) else ai_model

    # Try accessing openai api and checking if there is a model defined.
    # NOTE: This is not tested as no way to mock API currently.
    if api_keys and api_keys.openai:
        try:
            from openai import Client

            for model in Client(api_key=api_keys.openai).models.list().data:
                if model.id == name:
                    return True
        except ImportError:
            pass

    # Use the predefined list of models.
    return name in _ai_model_names


def _get_openai_model_max_tokens(name: str) -> int:
    """NOTE: OpenAI currently does not expose an API for getting the model
    length. Maybe add a config input value for this?"""

    # https://platform.openai.com/docs/models
    tokens = {
        "gpt-4o": 128000,
        "gpt-4": 8192,
        "gpt-3.5-turbo": 16385,
        "gpt-3.5-turbo-instruct": 4096,
    }

    # Split into - segments and remove each section from the end to find out
    # which one matches the most.

    # Base Case
    if name in tokens:
        return tokens[name]

    # Step Case
    name_split: list[str] = name.split("-")
    for i in range(1, name.count("-")):
        subname: str = "-".join(name_split[:-i])
        if subname in tokens:
            return tokens[subname]

    raise ValueError(f"Could not figure out max tokens for model: {name}")


def get_ai_model_by_name(
    name: str, api_keys: Optional[APIKeyCollection] = None
) -> AIModel:
    # Check OpenAI models.
    if api_keys and api_keys.openai:
        try:
            from openai import Client

            for model in Client(api_key=api_keys.openai).models.list().data:
                if model.id == name:
                    add_custom_ai_model(
                        AIModelOpenAI(
                            model.id,
                            _get_openai_model_max_tokens(model.id),
                        ),
                    )
        except ImportError:
            pass

    # Check AIModels enum.
    for enum_value in _AIModels:
        ai_model: AIModel = enum_value.value
        if name == ai_model.name:
            return ai_model

    # Check custom AI models.
    for custom_ai in _custom_ai_models:
        if name == custom_ai.name:
            return custom_ai

    raise Exception(f'The AI "{name}" was not found...')
