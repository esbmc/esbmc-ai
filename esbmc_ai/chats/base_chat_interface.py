# Author: Yiannis Charalambous

"""Contains code for the base class for interacting with the LLMs in a 
conversation-based way."""

from time import sleep, time
from typing import Any, Optional

from langchain.schema import (
    BaseMessage,
    HumanMessage,
    PromptValue,
)
from langchain_core.language_models import BaseChatModel
import structlog

from esbmc_ai.chat_response import ChatResponse, FinishReason
from esbmc_ai.ai_models import AIModel
from esbmc_ai.log_utils import LogCategories


class BaseChatInterface:
    """Base class for interacting with an LLM. It allows for interactions with
    text generation LLMs and also chat LLMs."""

    _last_attempt: float = 0
    cooldown_total: float = 20.0

    def __init__(
        self,
        system_messages: list[BaseMessage],
        ai_model: AIModel,
    ) -> None:
        super().__init__()
        self._logger: structlog.stdlib.BoundLogger = structlog.get_logger(
            category=LogCategories.CHAT
        )
        self.ai_model: AIModel = ai_model
        self._system_messages: list[BaseMessage] = system_messages
        self.messages: list[BaseMessage] = []

    def compress_message_stack(self) -> None:
        """Compress the message stack, is abstract and needs to be implemented."""
        self.messages = []

    def push_to_message_stack(
        self,
        message: BaseMessage | tuple[BaseMessage, ...] | list[BaseMessage],
    ) -> None:
        """Pushes a message(s) to the message stack without querying the LLM."""
        if isinstance(message, list) or isinstance(message, tuple):
            self.messages.extend(list(message))
        else:
            self.messages.append(message)

    def get_canonical_template_keys(
        self, source_code: str, esbmc_output: str, error_line: str, error_type: str
    ) -> dict[str, Any]:
        """Gets the canonical template keys for applying in template values."""
        return {
            "source_code": source_code,
            "esbmc_output": esbmc_output,
            "error_line": error_line,
            "error_type": error_type,
        }

    def apply_template_value(self, **kwargs: str) -> None:
        """Will substitute an f-string in the message stack and system messages to
        the provided value. The new substituted messages will become the new
        message stack, so the substitution is permanent."""

        system_message_prompts: PromptValue = self.ai_model.apply_chat_template(
            messages=self._system_messages,
            **kwargs,
        )
        self._system_messages = system_message_prompts.to_messages()

        message_prompts: PromptValue = self.ai_model.apply_chat_template(
            messages=self.messages,
            **kwargs,
        )
        self.messages = message_prompts.to_messages()

    def get_applied_messages(self, **kwargs: str) -> tuple[BaseMessage, ...]:
        """Applies the f-string substituion and returns the result instead of assigning
        it to the message stack."""
        message_prompts: PromptValue = self.ai_model.apply_chat_template(
            messages=self.messages,
            **kwargs,
        )
        return tuple(message_prompts.to_messages())

    def get_applied_system_messages(self, **kwargs: str) -> tuple[BaseMessage, ...]:
        """Same as `get_applied_messages` but for system messages."""
        message_prompts: PromptValue = self.ai_model.apply_chat_template(
            messages=self._system_messages,
            **kwargs,
        )
        return tuple(message_prompts.to_messages())

    @staticmethod
    def send_messages(
        ai_model: AIModel,
        messages: list[BaseMessage],
        logger: structlog.stdlib.BoundLogger | None = None,
    ) -> ChatResponse:
        """Static method to send messages."""

        # Check cooldown
        time_passed: float = time() - BaseChatInterface._last_attempt
        if time_passed < BaseChatInterface.cooldown_total:
            sleep_total_seconds: float = BaseChatInterface.cooldown_total - time_passed
            if logger:
                logger.info(f"Sleeping for {sleep_total_seconds}...")
            sleep(sleep_total_seconds)
        BaseChatInterface._last_attempt = time()

        response_message: BaseMessage = ai_model.invoke(input=messages)

        # Check if token limit has been exceeded.
        new_tokens: int = ai_model.get_num_tokens_from_messages(
            messages=messages + [response_message],
        )

        response: ChatResponse
        if new_tokens > ai_model.tokens:
            response = ChatResponse(
                finish_reason=FinishReason.length,
                message=response_message,
                total_tokens=ai_model.tokens,
            )
        else:
            response = ChatResponse(
                finish_reason=FinishReason.stop,
                message=response_message,
                total_tokens=new_tokens,
            )

        return response

    def send_message(self, message: Optional[str] = None) -> ChatResponse:
        """Sends a message to the AI model. Returns solution."""
        if message:
            self.push_to_message_stack(message=HumanMessage(content=message))

        all_messages = self._system_messages.copy()
        all_messages.extend(self.messages.copy())

        if message:
            self._logger.debug(f"LLM Prompt: {message}")

        response: ChatResponse = self.send_messages(
            ai_model=self.ai_model,
            messages=all_messages,
            logger=self._logger,
        )

        self._logger.debug(f"LLM Response: {message}")

        self.push_to_message_stack(message=response.message)
        return response
