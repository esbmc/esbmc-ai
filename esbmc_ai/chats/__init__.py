# Author: Yiannis Charalambous

"""This module contains different chat interfaces. Along with `BaseChatInterface`
that provides necessary boilet-plate for implementing an LLM based chat."""

from .base_chat_interface import BaseChatInterface
from .latest_state_solution_generator import LatestStateSolutionGenerator
from .solution_generator import SolutionGenerator
from .user_chat import UserChat

__all__ = [
    "BaseChatInterface",
    "LatestStateSolutionGenerator",
    "SolutionGenerator",
    "UserChat",
]
