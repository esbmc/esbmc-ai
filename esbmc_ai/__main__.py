#!/usr/bin/env python3

# Author: Yiannis Charalambous 2023

import os
from pathlib import Path
import re
import sys

# Enables arrow key functionality for input(). Do not remove import.
import readline
from typing import Optional

_ = readline

import argparse
from langchain.base_language import BaseLanguageModel


import esbmc_ai.config as config
from esbmc_ai import __author__, __version__
from esbmc_ai.solution import SourceFile, Solution, get_solution, init_solution

from esbmc_ai.commands import (
    ChatCommand,
    FixCodeCommand,
    HelpCommand,
    ExitCommand,
)

from esbmc_ai.loading_widget import LoadingWidget, create_loading_widget
from esbmc_ai.chats import UserChat
from esbmc_ai.logging import print_horizontal_line, printv, printvv
from esbmc_ai.esbmc_util import esbmc
from esbmc_ai.chat_response import FinishReason, ChatResponse
from esbmc_ai.ai_models import _ai_model_names


commands: list[ChatCommand] = []
command_names: list[str]
help_command: HelpCommand = HelpCommand()
fix_code_command: FixCodeCommand = FixCodeCommand()
exit_command: ExitCommand = ExitCommand()

chat: UserChat

HELP_MESSAGE: str = """Tool that passes ESBMC output into ChatGPT and allows for natural language
explanations. Type /help in order to view available commands."""

ESBMC_HELP: str = """Additional ESBMC Arguments - The following are useful
arguments that can be added after the filename to modify ESBMC functionality.
For all the options, run ESBMC with -h as a parameter:

  --compact-trace                  add trace information to output
  --no-assertions                  ignore assertions
  --no-bounds-check                do not do array bounds check
  --no-div-by-zero-check           do not do division by zero check
  --no-pointer-check               do not do pointer check
  --no-align-check                 do not check pointer alignment
  --no-pointer-relation-check      do not check pointer relations
  --no-unlimited-scanf-check       do not do overflow check for scanf/fscanf
                                   with unlimited character width.
  --nan-check                      check floating-point for NaN
  --memory-leak-check              enable memory leak check
  --overflow-check                 enable arithmetic over- and underflow check
  --ub-shift-check                 enable undefined behaviour check on shift
                                   operations
  --struct-fields-check            enable over-sized read checks for struct
                                   fields
  --deadlock-check                 enable global and local deadlock check with
                                   mutex
  --data-races-check               enable data races check
  --lock-order-check               enable for lock acquisition ordering check
  --atomicity-check                enable atomicity check at visible
                                   assignments
  --stack-limit bits (=-1)         check if stack limit is respected
  --error-label label              check if label is unreachable
  --force-malloc-success           do not check for malloc/new failure
"""


def check_health() -> None:
    printv("Performing health check...")
    # Check that ESBMC exists.
    if os.path.exists(config.esbmc_path):
        printv("ESBMC has been located")
    else:
        print(f"Error: ESBMC could not be found in {config.esbmc_path}")
        sys.exit(3)


def print_assistant_response(
    chat: UserChat,
    response: ChatResponse,
    hide_stats: bool = False,
) -> None:
    print(f"{response.message.type}: {response.message.content}\n\n")

    if not hide_stats:
        print(
            "Stats:",
            f"total tokens: {response.total_tokens},",
            f"max tokens: {chat.ai_model.tokens}",
            f"finish reason: {response.finish_reason}",
        )


def init_commands_list() -> None:
    # Setup Help command and commands list.
    global help_command
    commands.extend(
        [
            help_command,
            exit_command,
            fix_code_command,
        ]
    )
    help_command.set_commands(commands)

    global command_names
    command_names = [command.command_name for command in commands]


def update_solution(source_code: str) -> None:
    get_solution().files[0].update_content(content=source_code, reset_changes=True)


def _run_esbmc(source_file: SourceFile, anim: Optional[LoadingWidget] = None) -> str:
    assert source_file.file_path

    if anim:
        anim.start("ESBMC is processing... Please Wait")
    exit_code, esbmc_output = esbmc(
        path=source_file.file_path,
        esbmc_params=config.esbmc_params,
        timeout=config.verifier_timeout,
    )
    if anim:
        anim.stop()

    # ESBMC will output 0 for verification success and 1 for verification
    # failed, if anything else gets thrown, it's an ESBMC error.
    if not config.allow_successful and exit_code == 0:
        printv("Success!")
        print(esbmc_output)
        sys.exit(0)
    elif exit_code != 0 and exit_code != 1:
        printv(f"ESBMC exit code: {exit_code}")
        printv(f"ESBMC Output:\n\n{esbmc_output}")
        sys.exit(1)

    return esbmc_output


def init_commands() -> None:
    """# Bus Signals
    Function that handles initializing commands. Each command needs to be added
    into the commands array in order for the command to register to be called by
    the user and also register in the help system."""

    # Let the AI model know about the corrected code.
    fix_code_command.on_solution_signal.add_listener(chat.set_solution)
    fix_code_command.on_solution_signal.add_listener(update_solution)


def _run_command_mode(command: ChatCommand, args: argparse.Namespace) -> None:
    path_arg: Path = Path(args.filename)

    solution: Solution
    if path_arg.is_dir():
        solution = init_solution(
            Solution(
                [path for path in path_arg.glob("**/*") if path.is_file() and path.name]
            )
        )
    else:
        solution = init_solution(Solution([path_arg]))

    match command.command_name:
        case fix_code_command.command_name:
            for source_file in solution.files:
                # Run ESBMC first round
                esbmc_output: str = _run_esbmc(source_file)
                source_file.assign_verifier_output(esbmc_output)

                error, source_file_solution = fix_code_command.execute(
                    source_file=source_file
                )

                if error:
                    print("Failed all attempts...")
                else:
                    print(source_file_solution)
        case _:
            command.execute()
    sys.exit(0)


def parse_command(user_prompt_string: str) -> tuple[str, list[str]]:
    """Parses a command and returns it based on the command rules outlined in
    the wiki: https://github.com/Yiannis128/esbmc-ai/wiki/User-Chat-Mode"""
    regex_pattern: str = (
        r'\s+(?=(?:[^\\"]*(?:\\.[^\\"]*)*)$)|(?:(?<!\\)".*?(?<!\\)")|(?:\\.)+|\S+'
    )
    segments: list[str] = re.findall(regex_pattern, user_prompt_string)
    parsed_array: list[str] = [segment for segment in segments if segment != " "]
    # Remove all empty spaces.
    command: str = parsed_array[0]
    command_args: list[str] = parsed_array[1:]
    return command, command_args


def main() -> None:
    init_commands_list()

    parser = argparse.ArgumentParser(
        prog="ESBMC-ChatGPT",
        description=HELP_MESSAGE,
        # argparse.RawDescriptionHelpFormatter allows the ESBMC_HELP to display
        # properly.
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"Made by {__author__}\n\n{ESBMC_HELP}",
    )

    parser.add_argument(
        "filename",
        help="The filename to pass to esbmc.",
    )

    parser.add_argument(
        "remaining",
        nargs=argparse.REMAINDER,
        help="Any arguments after the filename will be passed to ESBMC as parameters.",
    )

    parser.add_argument(
        "-V",
        "--version",
        action="version",
        version="%(prog)s {version}".format(version=__version__),
        help="Show version information.",
    )

    parser.add_argument(
        "-v",
        "--verbose",
        action="count",
        default=0,
        help="Show up to 3 levels of verbose output. Level 1: extra information. Level 2: show failed generations, show ESBMC output. Level 3: print hidden pushes to the message stack.",
    )

    parser.add_argument(
        "-m",
        "--ai-model",
        default="",
        help="Which AI model to use. Built-in models: {OpenAI GPT models, "
        + ", ".join(_ai_model_names)
        + ", +custom models}",
    )

    parser.add_argument(
        "-r",
        "--raw-conversation",
        action="store_true",
        default=False,
        help="Show the raw conversation at the end of a command. Good for debugging...",
    )

    parser.add_argument(
        "-a",
        "--append",
        action="store_true",
        default=False,
        help="Any ESBMC parameters passed after the file name will be appended to the ones set in the config file, or the default ones if config file options are not set.",
    )

    parser.add_argument(
        "-c",
        "--command",
        choices=command_names,
        metavar="",
        help="Runs the program in command mode, it will exit after the command ends with an exit code. Options: {"
        + ", ".join(command_names)
        + "}",
    )

    args: argparse.Namespace = parser.parse_args()

    print(f"ESBMC-AI {__version__}")
    print(f"Made by {__author__}")
    print()

    config.load_envs()
    config.load_config(config.cfg_path)
    config.load_args(args)

    check_health()

    printv(f"Source code format: {config.source_code_format}")
    printv(f"ESBMC output type: {config.esbmc_output_type}")

    anim: LoadingWidget = create_loading_widget()

    # Read the source code and esbmc output.
    printv("Reading source code...")
    print(f"Running ESBMC with {config.esbmc_params}\n")

    assert isinstance(args.filename, str)

    # Command mode: Check if command is called and call it.
    # If not, then continue to user mode.
    if args.command != None:
        command = args.command
        if command in command_names:
            print("Running Command:", command)
            for idx, command_name in enumerate(command_names):
                if command == command_name:
                    _run_command_mode(command=commands[idx], args=args)
            sys.exit(0)

    # ===========================================
    # User Mode (Supports only 1 file)
    # ===========================================

    # Init Solution
    solution: Solution
    # Determine if we are processing one file versus multiple files
    path_arg: Path = Path(args.filename)
    if path_arg.is_dir():
        # Load only files.
        print(
            "Processing multiple files is not supported in User Mode."
            "Call a command using -c to process directories"
        )
        sys.exit(1)
    else:
        # Add the main source file to the solution explorer.
        solution = init_solution(Solution([path_arg]))
    del path_arg

    # Assert that we have one file and one filepath
    assert len(solution.files) == 1

    source_file: SourceFile = solution.files[0]
    assert source_file.file_path

    esbmc_output: str = _run_esbmc(source_file, anim)

    # Print verbose lvl 2
    print_horizontal_line(2)
    printvv(esbmc_output)
    print_horizontal_line(2)

    source_file.assign_verifier_output(esbmc_output)
    del esbmc_output

    printv(f"Initializing the LLM: {config.ai_model.name}\n")
    chat_llm: BaseLanguageModel = config.ai_model.create_llm(
        api_keys=config.api_keys,
        temperature=config.chat_prompt_user_mode.temperature,
        requests_max_tries=config.requests_max_tries,
        requests_timeout=config.requests_timeout,
    )

    printv("Creating user chat")
    global chat
    chat = UserChat(
        ai_model_agent=config.chat_prompt_user_mode,
        ai_model=config.ai_model,
        llm=chat_llm,
        source_code=source_file.latest_content,
        esbmc_output=source_file.latest_verifier_output,
        set_solution_messages=config.chat_prompt_user_mode.scenarios["set_solution"],
    )

    printv("Initializing commands...")
    init_commands()

    # Show the initial output.
    response: ChatResponse
    if len(config.chat_prompt_user_mode.initial_prompt) > 0:
        printv("Using initial prompt from file...\n")
        anim.start("Model is parsing ESBMC output... Please Wait")
        response = chat.send_message(
            message=config.chat_prompt_user_mode.initial_prompt,
        )
        anim.stop()

        if response.finish_reason == FinishReason.length:
            raise RuntimeError(f"The token length is too large: {chat.ai_model.tokens}")
    else:
        raise RuntimeError("User mode initial prompt not found in config.")

    print_assistant_response(chat, response)
    print(
        "ESBMC-AI: Type '/help' to view the available in-chat commands, along",
        "with useful prompts to ask the AI model...",
    )

    while True:
        # Get user input.
        user_message = input("user>: ")

        # Check if it is a command, if not, then pass it to the chat interface.
        if user_message.startswith("/"):
            command, command_args = parse_command(user_message)
            command = command[1:]  # Remove the /
            if command == fix_code_command.command_name:
                # Fix Code command
                print()
                print("ESBMC-AI will generate a fix for the code...")

                error: bool
                source_file_solution: str
                error, source_file_solution = fix_code_command.execute(
                    source_file=source_file
                )

                if not error:
                    print(
                        "\n\nESBMC-AI: Here is the corrected code, verified with ESBMC:"
                    )
                    print(f"```\n{source_file_solution}\n```")
                continue
            else:
                # Commands without parameters or returns are handled automatically.
                found: bool = False
                for cmd in commands:
                    if cmd.command_name == command:
                        found = True
                        cmd.execute()
                        break

                if not found:
                    print("Error: Unknown command...")
                continue
        elif user_message == "":
            continue
        else:
            print()

        # User chat mode send and process current message response.
        while True:
            # Send user message to AI model and process.
            anim.start("Generating response... Please Wait")
            response = chat.send_message(user_message)
            anim.stop()
            if response.finish_reason == FinishReason.stop:
                break
            elif response.finish_reason == FinishReason.length:
                anim.start(
                    "Message stack limit reached. Shortening message stack... Please Wait"
                )
                chat.compress_message_stack()
                anim.stop()
                continue
            else:
                raise NotImplementedError(
                    f"User Chat Mode: Finish Reason: {response.finish_reason}"
                )

        print_assistant_response(
            chat,
            response,
        )


if __name__ == "__main__":
    main()
