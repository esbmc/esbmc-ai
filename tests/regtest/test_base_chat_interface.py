# Author: Yiannis Charalambous

from langchain_core.language_models import FakeListChatModel
import pytest

from langchain.schema import BaseMessage, HumanMessage, AIMessage, SystemMessage

from esbmc_ai.ai_models import AIModel
from esbmc_ai.chat_response import ChatResponse
from esbmc_ai.chats.base_chat_interface import BaseChatInterface


@pytest.fixture
def setup():
    responses: list[str] = ["OK 1", "OK 2", "OK 3"]
    llm: FakeListChatModel = FakeListChatModel(responses=responses)

    ai_model: AIModel = AIModel("test", 1024)

    system_messages = [
        SystemMessage(content="System message"),
        AIMessage(content="OK"),
    ]

    chat: BaseChatInterface = BaseChatInterface(
        system_messages=system_messages,
        ai_model=ai_model,
        llm=llm,
    )

    return chat


def test_push_message_stack(regtest, setup) -> None:
    chat = setup

    messages: list[BaseMessage] = [
        AIMessage(content="Test 1"),
        HumanMessage(content="Test 2"),
        SystemMessage(content="Test 3"),
    ]

    chat.push_to_message_stack(messages[0])
    chat.push_to_message_stack(messages[1])
    chat.push_to_message_stack(messages[2])

    with regtest:
        for msg in chat._system_messages:
            print(f"{msg.type}: {msg.content}")

        for msg in chat.messages:
            print(f"{msg.type}: {msg.content}")


def test_send_message(regtest, setup) -> None:
    chat = setup

    chat_responses: list[ChatResponse] = [
        chat.send_message("Test 1"),
        chat.send_message("Test 2"),
        chat.send_message("Test 3"),
    ]

    with regtest:
        print("System Messages:")
        for m in chat._system_messages:
            print(f"{m.type}: {m.content}")
        print("Chat Messages:")
        for m in chat.messages:
            print(f"{m.type}: {m.content}")
        print("Responses:")
        for m in chat_responses:
            print(
                f"{m.message.type}({m.total_tokens} - {m.finish_reason}): {m.message.content}"
            )
