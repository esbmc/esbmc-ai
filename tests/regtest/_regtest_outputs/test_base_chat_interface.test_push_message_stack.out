system: System message
ai: OK
ai: Test 1
human: Test 2
system: Test 3
